tools/codegen/dest/ufunc.py

```diff
@@ -0,0 +1,440 @@
+    body += f"""
+else {{
+  gpu_kernel(iter, {ufunctor_sig.name}<scalar_t>({ufunctor_ctor_exprs_str}));
+}}
+    """
+    return body
+
+@with_native_function
+def compute_ufunc_cuda(g: NativeFunctionsGroup) -> str:
+    # First, build the functors, indexing them by dtype
+    ufunctor_sigs, ufunctors = compute_ufunc_cuda_functors(g)
+
+    # Next, build the conditionals
+    sig = StructuredImplSignature(g, ufunc.kernel_name(g, DispatchKey.CUDA))
+    dtype_cases = []
+    for dtype, inner_ufunctor_sigs in ufunctor_sigs.items():
+        dtype_cases.append(f"""
+AT_PRIVATE_CASE_TYPE("{sig.name}", at::ScalarType::{dtype}, {ScalarTypeToCppMapping[dtype]},
+  [&]() {{
+    {compute_ufunc_cuda_dtype_body(g, dtype, inner_ufunctor_sigs, sig.arguments())}
+  }}
+)
+""")
+
+    dtype_cases_str = "\n".join(dtype_cases)
+
+    return f"""
+{ufunctors}
+
+{sig.defn()} {{
+  TensorIteratorBase& iter = *this;
+  at::ScalarType st = iter.common_dtype();
+  RECORD_KERNEL_FUNCTION_DTYPE("{sig.name}", st);
+  switch (st) {{
+    {dtype_cases_str}
+    default:
+      TORCH_CHECK(false, "{sig.name}", " not implemented for '", toString(st), "'");
+  }}
+}}
+"""
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#
+#                                   CPU STUFF
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+@dataclass(frozen=True)
+class StubSignature:
+    g: NativeFunctionsGroup
+
+    @property
+    def name(self) -> str:
+        return f"{str(self.g.functional.func.name.name)}_stub"
+
+    @property
+    def kernel_name(self) -> str:
+        return f"{str(self.g.functional.func.name.name)}_kernel"
+
+    @property
+    def type_name(self) -> str:
+        return f"{str(self.g.functional.func.name.name)}_fn"
+
+    def arguments(self) -> List[Binding]:
+        return ufunc.stub_arguments(self.g)
+
+    def type(self) -> str:
+        cpp_args = self.arguments()
+        return f"void(*)(TensorIteratorBase&, {', '.join(a.type for a in cpp_args)})"
+
+    def dispatch_decl(self) -> str:
+        return f"DECLARE_DISPATCH({self.type_name}, {self.name})"
+
+    def dispatch_defn(self) -> str:
+        return f"DEFINE_DISPATCH({self.name})"
+
+    def kernel_defn(self) -> str:
+        return f"void {self.kernel_name}(TensorIteratorBase& iter, {', '.join(a.defn() for a in self.arguments())})"
+
+    def type_defn(self) -> str:
+        return f"using {self.type_name} = {self.type()}";
+
+    # must be called from context where this is TensorIteratorBase*
+    def call(self, ctx: Sequence[Binding]) -> str:
+        return f"{self.name}(device_type(), *this, {', '.join(a.expr for a in translate(ctx, self.arguments()))})"
+
+@with_native_function
+def compute_ufunc_cpu(g: NativeFunctionsGroup) -> str:
+    stub_sig = StubSignature(g)
+    sig = StructuredImplSignature(g, ufunc.kernel_name(g, DispatchKey.CPU))
```

Comment:
also minor point (I don't think a refactor is necessary, this is mostly just me thinking through the changes) - we need to know to use `ufunc.kernel_name(...)` both here where the ufunc kernel is defined, and later in `model.py` to know how the kernel is named in the `BackendIndex` (which is how we know what function to register to the dispatcher later).

I could imagine maybe writing this instead to grab the name from the `BackendIndex` instead (like we do in [RegisterDispatchKey](https://github.com/pytorch/pytorch/blob/7e35922a2cbc2eacfd9a11967d76db334b5b78d4/tools/codegen/dest/register_dispatch_key.py#L166)).


I guess the main difference here though is that we know here that we only want to support CPU and CUDA - and CPU/CUDA codegen also looks really different. Versus the `RegisterDispatchKey` codegen really is meant to be backend-agnostic, and try to abstract away all of the backend differences into the `BackendIndex` object (which we can't really do here).

Reply:
oh just saw the comment :) https://github.com/pytorch/pytorch/pull/65851/files?diff=unified&w=0#diff-7b8545c79e1049e63ea636145dbdaaf939208b02470357e99f4fb818c99be641R22